File: tools/skynet-client/lockstep/Core/RollingAverage.cs
namespace Skynet.DotNetClient.LockStep
{
    public class RollingAverage
    {
        public readonly int[] currentValues; //used only for logging
	
        private readonly int[] playerAverages;
        
        public RollingAverage(int numofPlayers, int initValue) {
            playerAverages = new int[numofPlayers];
            currentValues = new int[numofPlayers];
            for(int i=0; i<numofPlayers; i++) {
                playerAverages[i] = initValue;
                currentValues[i] = initValue;
            }
        }
	
        public void Add(int newValue, int playerId) {
            if(newValue > playerAverages[playerId]) {
                //rise quickly
                playerAverages[playerId] = newValue;
            } else {
                //slowly fall down
                playerAverages[playerId] = (playerAverages[playerId] * (9) + newValue * (1)) / 10;
            }
		
            currentValues[playerId] = newValue;
        }
	
        public int GetMax() {
            int max = int.MinValue;
            foreach(int average in playerAverages) {
                if(average > max) {
                    max = average;
                }
            }
		
            return max;
        }
    }
}

File: service/battle/lockstep/core/rolling_average.lua
local class	   =  require "class"
local RollingAverage   = class("RollingAverage")


function RollingAverage:initialize(data)
    self.currentValues = array_new(data.NumberOfPlayers, 0)
    self.playerAverages = array_new(data.NumberOfPlayers, 0)
    for i=1, data.NumberOfPlayers do
        self.playerAverages[i] = data.initValue
        self.currentValues[i] = data.initValue
    end
end

function RollingAverage:Add(newValue, playerId)
    if newValue > self.playerAverages[playerId] then
        -- rise quickly
        self.playerAverages[playerId] = newValue
    else
        -- slowly fall down
        self.playerAverages[playerId] = (self.playerAverages[playerId] * (9) + newValue * (1)) / 10
    end

    self.currentValues[playerId] = newValue
end

function RollingAverage:GetMax()
    local max = 0
    for i=1, #self.playerAverages do
        local average = self.playerAverages[i]
        if average > max then
            max = average
        end
    end
    return max
end


return RollingAverage